### ISSUE TITLE
Git-style 3 dot diff?

### ISSUE THREAD

**Issue by addcninblue** [jj-vcs/jj#4915]
Git has a 3-dot diff, which shows diffs introduced by a branch. I believe it knows to omit changes introduced due to a merge commit.

[Github PRs use a three-dot merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-comparing-branches-in-pull-requests#:~:text=Pull%20requests%20on%20GitHub%20show%20a%20three%2Ddot%20diff). I'd ideally like to be able to reproduce this type of diff locally so I can reason about the change.

In Jujutsu, I'm not sure this feature is easy or even possible. For feature branches without merges, I think using `fork_point` (https://github.com/martinvonz/jj/pull/4795) is sufficient (as one could get the `fork_point` between the head of a branch and trunk). However, for branches with merge commits, it doesn't seem currently possible. (Furthermore, when I think about Jujutsu concepts, it seems like it's possible it wasn't meant to handle this type of thinking.)

Question: Is my way of thinking wrong for Jujutsu? Or is there some way to get this feature via `jj diff` and revsets somehow? Thanks in advance!

**Comment by arxanas** [comment]
By the way, since it might be relevant for this discussion, I'll just note now that `git diff`'s interpretation of three dots is specifically different from every other command's interpretation of three dots (which normally means "symmetric difference": https://git-scm.com/docs/gitrevisions#Documentation/gitrevisions.txt-Theememthree-dotSymmetricDifferenceNotation).

**Comment by martinvonz** [comment]
Yes, it's very confusing. And `A..B` should have the meaning that `A...B` does but instead it simply means `A B` :(

**Comment by martinvonz** [comment]
Btw, a related idea that came up on IRC not too long ago is that `jj diff -r` (which currently supports a single commit) could be generalized to allow a revset by showing the cumulative differences of the commits in the set. If you pass it `jj diff -r A..B` would would then get the `git diff A...B` behavior.

**Comment by addcninblue** [comment]
Wow, thanks for the quick replies! Let me play around with the suggestions and get a better understanding before I comment. I think martinvonz's comment above is accurate (that adding revset support to diffs is sufficient to get a proper git-style 3 dot diff)

**Comment by martinvonz** [comment]
> Would an implementation of `jj diff -r '@ | revision'` do sorting in this order automatically?

If we make `jj diff -r` support multiple revisions, we might make it error out if the sets are discontiguous. Or we might make it temporarily rebase the commits so they become contiguous if they weren't. (`jj interdiff` and `jj evolog` do temporary rebasing, but the rebase destination is less ambiguous in those cases.)

**Comment by ilyagr** [comment]
> Btw, a related idea that came up on IRC not too long ago is that `jj diff -r` (which currently supports a single commit) could be generalized to allow a revset by showing the cumulative differences of the commits in the set. If you pass it `jj diff -r A..B` would would then get the `git diff A...B` behavior.

This is pretty clever.

> Or we might make it temporarily rebase the commits so they become contiguous if they weren't.

I think this actually makes sense in theory (though it is quite confusing, and I'm ignoring the A+A-B=A problem which breaks this somewhat). For `jj diff -r X`, you could take the auto-merge of `roots(X)-`, rebase each commit in `X` on top of that (individually), and then take the auto-merge of all those rebased commits. You can then show the diff of the two auto-merges.

The issue of auto-merging the parents of roots will come up even if the set is contiguous, as long as it has multiple roots. (Which we could also instead prohibit, for simplicity, of course).

**Comment by Dianne087** [comment]
references z

**Comment by kwiberg** [comment]
`jj diff -r` with multiple revisions sounds great! (It’s also one of the first things I tried before realizing that this feature didn’t already exist.)

It seems like there are a few cases to consider, based on the given revset:

1. **One base commit** (not in the revset, but parent to at least one commit in the revset), **one leaf commit** (in the revset, but has no children in the revset). Here, we just show the diff from the base commit to the leaf commit.
2. **Two or more base commits, one leaf commit.** If one of the base commits is a descendant of all the others, we show the diff between that one and the leaf commit. Otherwise, we complain that the diff isn’t well-defined.
3. **Two or more leaf commits.** No matter how many base commits there are, we complain that the diff isn’t well-defined.

Note that (1) is just a special case of (2). (3) will—among other things—disqualify all discontiguous revsets.

I think this covers all the cases where we can identify a unique “before” and a unique “after” revision, which is exactly the cases where we can show a unique diff. For example, `jj diff -r base..tip` would behave just as expected: show the diff from the most recent merge to `tip`, or complain if there isn’t a unique most recent merge.

**Comment by martinvonz** [comment]
There can also be cases like `jj diff -r 'A|C'` (skipping `B` in the middle). Perhaps this is your (2), but I don't think we should diff `A-` to `C` in this case because it would include the changes from `B`. I think we should either error out or do a temporary rebase of `C` onto `A` and diff from `A-` to that temporary commit. But I think a good start is to just disallow it :)

**Comment by kwiberg** [comment]
`A | C` would be case (3), because both A and C are leaves (are in the revset, but have no children in the revset). So my rules would disallow it, and any other discontiguous revset.

In other words, my rules and I agree with what you said!

**Comment by martinvonz** [comment]
Ah, makes sense :) What's an example of (2) then? I guess you can have something like this:

```
$ jj log
D
|\
C |
| |
B |
|/
A
$ jj diff -r 'A | C | D'
```

That set has one head (`D`) and two bases (`A` and `C`), and one base is a descendant of the other, but I don't think we want to show the diff from `C-` to `D` there, because that would miss the diff in `A`.

**Comment by kwiberg** [comment]
(In your example, the bases would be `A- | C-`, since the bases are the commits that aren’t in the revset, but are parents to at least one commit in the revset.)

Ah, yes, we must also require that the revset be convex (meaning that if X and Y are in the set, then `X::Y` are all in the set). Otherwise we have a situation where parts of the revset has already been merged (in the sense that those commits are ancestors of the base commits), and my rules would say that we should omit them from the diff, which would be confusing.

`A | C | D` is not convex (because it doesn’t include B), so we wouldn’t generate a diff for it.

An example of (2) with a properly convex revset is the following:

```
     b---c-------------0
    /     \             \
---a       f---main      my-branch
    \     /             /
     d---e-------------1
```

Here, `main..my-branch` contains three commits: 0, 1, and my-branch. There is just one leaf commit (my-branch), but two base commits (c and e), and neither is an ancestor of the other. In this situation, we have no good way to pick a unique base for the diff, so we should complain and give up.

Restating the algorithm with convexity included:

1’. If the revset doesn’t have exactly one leaf, complain and quit. (Because we need a unique end point for the diff.)
2’. If the revset doesn’t have a base that’s a descendant of all the other bases, complain and quit. (Because we need a unique starting point for the diff.)
3’. If the revset isn’t convex, complain and quit. (Because if there are holes in the revset, the diff won’t accurately describe the whole set.)
4’. Print the diff from the descendant base to the leaf.
